gips/0046-graphql-validations.md

```diff
@@ -0,0 +1,108 @@
+- `SingleFieldSubscriptions` (https://spec.graphql.org/draft/#sec-Operation-Name-Uniqueness)
+- `KnownTypeNames` (https://spec.graphql.org/draft/#sec-Fragment-Spread-Type-Existence)
+- `VariablesAreInputTypes` (https://spec.graphql.org/draft/#sec-Variables-Are-Input-Types)
+- `FieldsOnCorrectType` (https://spec.graphql.org/draft/#sec-Field-Selections)
+- `LoneAnonymousOperation` (https://spec.graphql.org/draft/#sec-Lone-Anonymous-Operation)
+- `FragmentsOnCompositeTypes` (https://spec.graphql.org/draft/#sec-Fragments-On-Composite-Types)
+- `OverlappingFieldsCanBeMerged` (https://spec.graphql.org/draft/#sec-Field-Selection-Merging)
+- `NoUnusedFragments` (https://spec.graphql.org/draft/#sec-Fragments-Must-Be-Used)
+- `KnownFragmentNamesRule` (https://spec.graphql.org/draft/#sec-Fragment-spread-target-defined)
+- `LeafFieldSelections` (https://spec.graphql.org/draft/#sec-Leaf-Field-Selections)
+- `UniqueFragmentNames` (https://spec.graphql.org/draft/#sec-Fragment-Name-Uniqueness)
+- `NoFragmentsCycle` (https://spec.graphql.org/draft/#sec-Fragment-spreads-must-not-form-cycles)
+- `PossibleFragmentSpreads` (https://spec.graphql.org/draft/#sec-Fragment-spread-is-possible)
+- `NoUnusedVariables` (https://spec.graphql.org/draft/#sec-All-Variables-Used)
+- `NoUndefinedVariables` (https://spec.graphql.org/draft/#sec-All-Variable-Uses-Defined)
+- `KnownArgumentNames` (https://spec.graphql.org/draft/#sec-Argument-Names , https://spec.graphql.org/draft/#sec-Directives-Are-In-Valid-Locations)
+- `UniqueArgumentNames` (https://spec.graphql.org/draft/#sec-Argument-Names)
+- `UniqueVariableNames` (https://spec.graphql.org/draft/#sec-Variable-Uniqueness)
+- `ProvidedRequiredArguments` (https://spec.graphql.org/draft/#sec-Required-Arguments)
+- `ValuesOfCorrectType` (https://spec.graphql.org/draft/#sec-Values-of-Correct-Type)
+- `VariablesInAllowedPosition` (https://spec.graphql.org/draft/#sec-All-Variable-Usages-Are-Allowed)
+- `KnownDirectives` (https://spec.graphql.org/draft/#sec-Directives-Are-Defined)
+- `UniqueDirectivesPerLocation` (https://spec.graphql.org/draft/#sec-Directives-Are-Unique-per-Location)
+
+The following spec rules are not part of this implementation:
+
+- `ExecutableDefinitions` - not needed, because of `graphql_parser` nature.
+- `UniqueInputFieldNames` (blocked by https://github.com/graphql-rust/graphql-parser/issues/59, low priority for `graph-node`)
+
+In case of a validation error as part of the `validate` phase, `graph-node` will return a validation error, and will skip the GraphQL execution:
+
+![GraphQL Validation Error](../assets/gip-0046/graphql-validations-1.png)
+
+![GraphQL Validation Error](../assets/gip-0046/graphql-validations-2.png)
+
+# Backwards Compatibility
+
+The GraphQL execution layer will run in the same way, but due to the nature of the extra added validation checks, some users might experience a change in the way `graph-node` responds to end-users GraphQL operations.
```

Comment:
I think it is worth explicitly stating that this will result in queries which returned successfully now returning errors in a newer version of Graph Node.